### ISSUE TITLE
Rotating masterkey

### ISSUE THREAD

**Issue by iFrozenPhoenix** [zitadel/zitadel#6768]
### Preflight Checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions
- [ ] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe your problem

I'd like to rotate the master key on a regular basis to keep the underlying secrets secure.

### Describe your ideal solution

If the key should be rotated the instance is started with the flags masterkey (contains the new key) and masterkey-old (with the old masterkey).
The old masterkey is then used to decrypt the secrets and the new one to encrypt them again. After the process finished the old masterkey cannot be used anymore. If the process is started again with these flags and masterkey is already the currently used masterkey nothing happens and the instance runs normally.

### Version

latest

### Environment

Self-hosted

**Comment by hifabienne** [comment]
Thank you for sharing your idea.
If there is a significant demand from customers/community, we will carefully consider implementing the feature.
Currently, the issue will be added to our product backlog.

Meanwhile, if you're interested in implementing it yourself, we also welcome pull requests.

**Comment by hifabienne** [comment]
@muhlemmer @livio-a is this implemented with the new webkey implementation? https://github.com/zitadel/zitadel/pull/8508

**Comment by muhlemmer** [comment]
No, this issue is about the masterkey used for encryption in the storage. Currently, if you loose or change the masterkey on the commandline, you essentially loose access to all stored secrets in zitadel.

**Comment by sandeep1699** [comment]
Is there a plan to implement this feature? It would greatly assist us in addressing security issues.

**Comment by muhlemmer** [comment]
> Is there a plan to implement this feature? It would greatly assist us in addressing security issues.

Not our priority. As zitadel is event driven it is rather complex rotating keys, as it requires re-encrypting some existing events. We currently don't have the resources to implement this.

**Comment by wizardist** [comment]
@muhlemmer what are those events? Why is masterkey encrypting those and not one of the secrets in the secret store?

The [Secrets article](https://zitadel.com/docs/concepts/architecture/secrets) claims the following:

> ZITADEL uses the following principles when handling Secrets across their lifecycle:
> - Automate rotation
> - ...

Does Zitadel rotate all secrets? Why doesn't it then affect the event store?

**Comment by Larzous** [comment]
@matiasracedo pointed me to this ticket due to my request about rotating master keys.

@hifabienne 
> If there is a significant demand from customers/community, we will carefully consider implementing the feature.

per @muhlemmer 
> Not our priority. As zitadel is event driven it is rather complex rotating keys, as it requires re-encrypting some existing events. We currently don't have the resources to implement this.

**My Take**
As a DevOps Engineer and server administrator, having the ability to rotate a `masterkey` is a **vital and fundamental** necessity for **security** for a few major events:
- Server Compromise
 - server gets hacked and key is in text file, bash_history, or docker configuration is now exposed
 - on server recover/restoration keys/passwords should an do get rotated for **security**.
- Employee leaving
  - it is recommended that anything that uses a static key gets rotated when key individuals who had access leave the company.
- compromised/exposed key
  - they key gets oopsied into something non-secure and is considered compromised.

**Reason**
The core reason for a feature like this, the reason we need proper password management, **SECURITY**!  Core principled password management and password rotation.  Within Zitadel we have the ability to force expire passwords after X days, but we don't have the ability to rotate the `masterkey` at the heart of the database?  Without the ability to rotate a master key, we would be forced to export every configuration and use, and import them into a new server/installation/setup of Zitadel.  This kind of unnecessary work using v1 APIs is a step that quite honestly could be turning people away from using Zitadel.

**Recommendation**
As part of a `setup` phase, allow a new flags:
- `--new-masterkey <new-key>`
- `--new-masterkeyFile <file containing new-key>`

**Example**
```cmd
zitadel setup --config /home/zitadel/config.yaml --masterkeyFile /home/zitadel/zitadel-masterkey-old --new-masterkeyFile /home/zitadel/zitadel-masterkey
```

**Comment by muhlemmer** [comment]
@wizardist secrets are rotated and then stored encrypted using the masterkey. Primary secrets leaving the system, such as PATs, signing keys, client secrets to IDP etc, can be rotated for security purposes.

On top of that we store those secrets encrypted-at-rest using AES symmetric encryption. As Zitadel is also event-driven, all data modifications are stored in an append-only table. We build our data models from those events. However, in order for rotation to make sense, historical data must be re-encrypted upon rotation. This is something that breaks the append-only purpose of the event store. Therefore the masterkey itself can't be rotated, that's what this issue is about.

@Larzous The masterkey based encryption is used as an additional measure and a specific feature not available in all products. Passwords are always hashed, not encrypted using the masterkey so they comply to security best-practices already. Hash algorithms can also be rotated!

Note you can reconfigure the masterkey at any moment. Any stored secrets will instantly be invalid. Some secrets are used to integrate with 3rd parties, like IDPs. You'll need to rotate the secrets at the provider and then set them accordingly in Zitadel. Some secrets like private keys will need to be re-generated.

Now, in order to "exploit" a long living masterkey, an attacker needs to:
1. Have direct access to the database. Or restored a backup dump.
2. Have access to the runtime secrets to read the masterkey

Let's say you became aware of a leak, wouldn't you still rotate the primary secrets? They may have already been read by the attacker.

Case 2: off-boarding an employee. If they maliciously keep a copy of the database somewhere (pre-req 1) and have the old masterkey, rotation of the masterkey won't help you. They would still gain access to the primary secrets. You **still** need to rotate all of them.

IMO the best measure is to keep short-lived primary secrets. There **_may_** be additional security benefits to masterkey rotation, but they mean little against the above attack vector.

Because of implementation complexities in the _current_ architecture and little benefit, this issue is not a priority for us.   
 
After we achieve https://github.com/zitadel/zitadel/issues/9599 we should be able to re-evaluate, because payload data of events becomes irrelevant for state. In other words, it would be easier to redact events and rewrite existing data in relational tables.

**Comment by wizardist** [comment]
Thanks for the explanation, @muhlemmer.

Have you considered key wrapping? Use master key as the KEK, generate a bunch of DEKs for different purposes (don't you do it already?) that you don't really need to rotate. Rotating the master key would re-wrap DEKs – no need to mutate the actual data ciphertexts. Works well for both event sourcing and relational storage models.

**Comment by fforootd** [comment]
What about a Key Ring (list of keys) approach?

It keeps things stateless and declarative, which is super important for K8s, but works for everyone.

**The Logic**

1. Config: Just accepts a list: MasterKeys: ["key-new", "key-old"].
2. Write: Always encrypt with the first key (key-new).
3. Read: Try the first key; if that fails, try the next one (key-old).

**How rotation would look... It basically turns rotation into a standard config update:**

1. Add Key: Update your config/secret to ["key-new", "key-old"] and redeploy (e.g., kubectl rollout restart). Result: App reads old data fine, but starts writing everything with the new key immediately.
2. Migrate: As data is accessed/saved, it gets re-encrypted with key-new. Optional: We could add a zitadel keys migrate CLI command to force this in the background if needed.
3. Cleanup: Once the old key is unused, update config to just ["key-new"] and redeploy.

This would also work with KMS systems easily without us needing to engineer full Envelope Encryption right now. Ops teams can just inject the key list from their provider (Vault, AWS, etc.) into the config at runtime. Keeps the complexity low for us while still letting them manage keys externally.

Thoughts? Esp. @mridang since we talked about this? (I know I had a different opinion in that chat 😁 )

**Comment by LTSlw** [comment]
I’ve built an “offline” masterkey rotation tool.  
https://codeberg.org/LTSlw/zitadel-masterkey-rotator
